core/templates/dev/head/services/CsrfTokenServiceSpec.ts

```diff
@@ -0,0 +1,59 @@
+// Copyright 2014 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for the csrf service
+ */
+
+require('services/CsrfTokenService.ts');
+
+describe('Csrf Token Service', function() {
+  var CsrfTokenService, $httpBackend;
+
+  beforeEach(angular.mock.module('oppia'));
+  beforeEach(angular.mock.inject(function($injector, $q) {
+    CsrfTokenService = $injector.get('CsrfTokenService');
+    $httpBackend = $injector.get('$httpBackend');
+
+    spyOn($, 'ajax').and.callFake(function() {
+      var deferred = $q.defer();
+      deferred.resolve('sample-csrf-token');
+      return deferred.promise;
+    });
+  }));
+
+  it('should correctly set the csrf token', function() {
+    CsrfTokenService.initializeToken();
+
+    CsrfTokenService.getTokenAsync().then(function(token) {
+      expect(token).toBe('sample-csrf-token');
+    });
+  });
+
+  it('should error if initialize is called more than once',
+    function() {
+      CsrfTokenService.initializeToken();
+
+      expect(CsrfTokenService.initializeToken).toThrowError(
+        'Token request already made');
+    }
+  );
+
+  it('should error if getTokenAsync is called before initialization',
```

Comment:
@seanlip  It is tested here.

Reply:
I called getTokenAsync without calling initialization

Reply:
Wait, I'm confused, sorry. You said earlier: 

> @seanlip I think we'll still need to return a promise because it is possible that the initialization is not complete when a POST or PUT request is sent so the request will have to wait for the initialization to be complete.

But this test is checking that it throws an error, which seems different from your description.

Also, where does "Token needs to be initialized" come from? I do not see it in core/templates/dev/head/services/CsrfTokenService.ts

Reply:
The test checks if getTokenAsync is called without already calling initializeToken.
It throws the error if a call to getTokenAsync was made before initializing the token